Propagate thread locals to Delta thread pools

fred-db opened the pull request:
#### Which Delta project/connector is this regarding?
<!--
Please add the component selected below to the beginning of the pull request title
For example: [Spark] Title of my pull request
-->

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

<!--
- Describe what this PR changes.
- Describe why we need the change.
 
If this PR resolves an issue be sure to include "Resolves #XXX" to correctly link and close the issue upon merge.
-->
* The default thread pool executor in Apache Spark does not forward thread locals to threads spawned in a thread pool.
*  This can cause issues if the threads depend on the thread locals.
* To fix this, we introduce a wrapper class around the thread pool executor that forwards thread locals.

## How was this patch tested?

<!--
If tests were added, say they were added here. Please make sure to test the changes thoroughly including negative and positive cases if possible.
If the changes were tested in any way other than unit tests, please clarify how you tested step by step (ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future).
If the changes were not tested, please explain why.
-->
* UTs of `SparkThreadLocalForwardingExecutor` to ensure thread locals are forwarded and reset after future finished.

## Does this PR introduce _any_ user-facing changes?

<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Delta Lake versions or within the unreleased branches such as master.
If no, write 'No'.
-->
No

larsk-db commented on spark/src/main/scala/org/apache/spark/sql/delta/util/threads/sparkThreadLocalForwardingThreadPoolExecutor.scala:
```diff
@@ -0,0 +1,132 @@
+  // Capture an immutable threadsafe snapshot of the current local properties
+  val capturedProperties = sparkContext
+    .map(sc => CapturedSparkThreadLocals.toValuesArray(
+      org.apache.spark.util.Utils.cloneProperties(sc.getLocalProperties)))
```
import `Utils` instead of using qualified name here? You can rename it (e.g. to `SparkUtils`) if it clashes.

larsk-db commented on spark/src/main/scala/org/apache/spark/sql/delta/util/threads/sparkThreadLocalForwardingThreadPoolExecutor.scala:
```diff
@@ -0,0 +1,132 @@
+    val previousProperties = sparkContext.map(_.getLocalProperties)
+
+    TaskContext.setTaskContext(capturedTaskContext)
+    capturedProperties.foreach { p =>
+      sparkContext.foreach(_.setLocalProperties(CapturedSparkThreadLocals.toProperties(p)))
+    }
```
```suggestion
    for {
      p <- capturedProperties
      sc <- sparkContext
    } sc.setLocalProperties(CapturedSparkThreadLocals.toProperties(p))
```
Isn't this much more readable? ;)

larsk-db commented on spark/src/main/scala/org/apache/spark/sql/delta/util/threads/sparkThreadLocalForwardingThreadPoolExecutor.scala:
```diff
@@ -0,0 +1,132 @@
+        throw t
+    } finally {
+      TaskContext.setTaskContext(previousTaskContext)
+      previousProperties.foreach(p => sparkContext.foreach(_.setLocalProperties(p)))
```
same here

larsk-db commented on spark/src/main/scala/org/apache/spark/sql/delta/util/threads/sparkThreadLocalForwardingThreadPoolExecutor.scala:
```diff
@@ -0,0 +1,132 @@
+
+  def toProperties(props: Array[(String, String)]): Properties = {
+    val resultProps = new Properties()
+    props.foreach { kvp =>
+      resultProps.put(kvp._1, kvp._2)
+    }
```
Or, hear me out:
```suggestion
    for ((key, value) <- props) {
      resultProps.put(key, value)
    }
```

larsk-db commented on spark/src/test/scala/org/apache/spark/sql/delta/util/SparkThreadLocalForwardingSuite.scala:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  test("That CapturedSparkThreadLocals properly restores the existing spark properties." +
+  " Changes to local properties inside a task do not affect the original properties") {
```
indent

larsk-db reviewed (commented):
Mostly nits, otherwise it's looks good thanks :)

fred-db commented:
Hi @larsk-db , thank you so much for reviewing! :) I addressed your comments, hope it looks good now. I also moved the `DeltaThreadPool` under `util/threads`, as we now have more threading-specific code in the codebase and it makes sense to put it into a separate package.

larsk-db reviewed (approved):
LGTM thank you!
